### DISCUSSION TITLE
Activating my extension only in certain markdown files

### DISCUSSION THREAD

**Discussion by ejgallego** [microsoft/vscode-discussions#901] [Extension Development QnA] [answered]
Dear VSCode community,

**summary**: we'd like our extension to activate only on markdown files with extension `.mv`

we have an extension (https://github.com/ejgallego/coq-lsp) for the Coq Interactive Theorem prover. This extension opens a side panel showing proof information files as it is customary in the Coq / Interactive Theorem Proving community.

Our extension also supports literate markdown documents, that is to say, markdown documents with embedded Coq code snippets marked with `coq` blocks.

Ideally, we'd like our extension to only activate for markdown files that have the `.mv` extension, as we think it'd be intrusive for our extension to activate on all markdown files, due to the right side panel. We'd also like for the primary type of our `.mv` files to stay as Markdown, as it makes sense due to preview, etc...

However, we are struggling having our extension to activate automatically when a .mv file is opened. As of now, if we open an .mv file, the extension won't activate. Once the extension is activated by other means, it works fine with the .mv files.

This is our contribution point:
```js
  "contributes": {
    "languages": [
      {
        "id": "markdown",
        "aliases": [
          "Markdown"
        ],
        "extensions": [
          ".mv"
        ]
      }
    ]
```

Is it possible for us to do what we want? 

TIA!

**Comment by Ninglo** [comment]
First, we can set up the extension [activation event](https://code.visualstudio.com/api/references/activation-events#workspaceContains):
```JSON
"activationEvents": [
    "workspaceContains:**/*.mv"
]
```

After the extension is activated, we can observe workspace event changes (for example: [onDidOpenTextDocument](https://code.visualstudio.com/api/references/vscode-api#workspace
)) and we can perform real logic when the user opens a specific file.
```JS
export function activate() {
  vscode.workspace.onDidOpenTextDocument((textDocument) => {
    if (
      textDocument.languageId === "markdown" &&
      textDocument.name.endsWith(".mv")
    ) {
      // Do something
    }
  });
}
```

Note that these codes have not been actually tested, so some content may not be completely accurate and are for reference only.

Hope this helps you.

**Reply by Ninglo** [reply]
One thing to add is that you don't need to execute all the logic when activating the plugin. You can just listen for events and respond when relevant events occur.

**Comment by alefragnani** [comment]
Hi @ejgallego ,

There is no single `activationEvent` setting to achieve what you want, but you still can use take advantage of the regular Markdown features (like the Preview you mentioned), if you add this simple setting:

```json
    "files.associations": {
        "*.mv": "markdown"
    },
```

Using `files.associations` setting, whenever you open a `.mv` file, it will be recognized as `markdown`.

I'm not entirely sure, but I guess that this may work even if your `.mv` file is defined as a _new language_. Sure, this would require your extension to update the user settings somehow (a command, instructions in readme, step in walkthrough) but it is doable, Worth a try 😬 

Hope this helps

**Comment by theReynald** [comment]
Hi @ejgallego,

Does the follow responses help?

**Comment by ejgallego** [comment]
Hi all! Thanks a lot for all the amazing answers and help, I will look into them this upcomign week, I was offline for the past week.

Something that has me wondering is that maybe there could be a bug in the new activation code.

With the current code, when I activate the extension manually, it is properly activated on all the .mv files, so what it fails is the auto-activation, but I don't see how this should be the case.

I'll test and report back soon, thanks for your patience.

**Comment by r3m0t** [comment, marked as answer]
Hi, I think `onLanguage:markdown` would be a better event for you. Also, once activated make sure to loop through all open editors for mv files. For example if the user restores a window from the last time they had VS Code open, the events may not fire or may have fired before your extension activated.

**Comment by ejgallego** [comment]
Hi folks, thanks again for all the help, but the more I think about this, the more it seems a bug (likely due to the new activation setup in 1.74).

It makes little sense to me that the following happens as of now:
- when the extension was not activated, `.mv` files are not recognized
- however when the extension was activated by other means, the extension is correctly active for old and new `.mv` files.

Isn't that strange? Dunno if I should file a bug.

> Hi, I think onLanguage:markdown would be a better event for you. Also, once activated make sure to loop through all open editors for mv files. For example if the user restores a window from the last time they had VS Code open, the events may not fire or may have fired before your extension activated.

Adding `onLanguage:markdown` plus my current `contributes` code makes things work as I want, thanks!

That would be a solution, however I am not sure this is not a bug; I'd like to determine if that looks like a bug before closing the question.

What do you mean by "the events may not fire or may have fired before your extension activated" ?

I'd like to add a test case for this, but I wasn't able to reproduce this behavior so I cannot test the code that would iterate the docs.

Note that my extension is a LSP client one, so once activated the LSP client takes care of synching the editor, etc...

**Reply by isidorn** [reply]
Might be best to file a bug here https://github.com/microsoft/vscode
And someone from the dev team should be able to help. Thank you

**Reply by ejgallego** [reply]
Thanks a lot @isidorn, indeed I think I have collected enough evidence to warrant bothering the devs with this. Thanks again to everyone!
